Break everything and introduce i18n

benoit74 opened the pull request:
## Rationale
- Introduce i18n of the UI and the emails
- Move ui to Vue 3, Vuetify, Pinia, Vue-Router
- Move api to FastAPI, hatch
- Fix many small glitches linked to recent Zimfarm changes
- Enhance some messages
- Upgrade to Node 20 and Python 3.12
- Adapted CI to check code QA on both api and ui + add automated tests
- Split API and UI for proper deployment (two Dockerfile, two packages, adapt CI, ... work on kiwix/operations will follow)

Nota: 
- for now PR contains fake and incomplete translations to fr and fa, but these will be dropped before merging, this is used only for testing purpose

rgaudin commented on ui/src/App.vue:
```diff
@@ -1,35 +1,97 @@
+    </v-select>
+    <RouterView />
+    <i18n-t keypath="footer.paragraph" tag="footer" class="text-center">
+      <a target="_blank" href="https://kiwix.org">{{ t('footer.link0') }}</a>
```
Same remark about keys naming

rgaudin reviewed (changes requested):
That's something! Thank you.

I tested it and it seems to work fine. I was somehow surprised that the UI looks the same 😀

On the python side, I think there are a couple of place with poor type hints (zimfarm module?).

On the Vue part, I mostly skimmed through the code but it looks good and coherent.

On personal taste, I find the user feedback too dim. If there is an error in making a request, theres just a light toast at the bottom of the page that disappears after 5s.
I had to make several requests to see it.

benoit74 commented on .github/workflows/Publish.yml:
```diff
@@ -22,8 +22,7 @@ jobs:
           registries: ghcr.io
-          credentials:
-            GHCRIO_USERNAME=${{ secrets.GHCR_USERNAME }}
+          credentials: GHCRIO_USERNAME=${{ secrets.GHCR_USERNAME }}
```
This is `prettier` automated formatting. I've added a disabling rule for this line.

benoit74 commented on .github/workflows/QA.yml:
```diff
@@ -7,24 +7,59 @@ on:
+      - name: Check eslint
+        working-directory: ui
+        run: |
+          yarn lint
```
Indeed

benoit74 commented on api/pyproject.toml:
```diff
@@ -0,0 +1,208 @@
+    "pydantic==2.9.1",       # this is also a sub-dep of fastapi but we rely a lot on it
+    "pyhumps==3.8.0",
+    "requests==2.32.3",
+    "python-i18n==0.3.9",
```
I completely forgot to check this, thank you.

I agree this is a concern, but:
- there is not really an alternative from my PoV, gettext / Babel seem to be really complex for our simple need, and not really compatible with JSON
- there are already forks available, maybe one will achieve to merge one day or the other
- the codebase is really very limited, shouldn't be a big concern to maintain on our own should things go wild

Anyway I've quite isolated the translation module so that we can easily switch from one to another if needed.

I propose to continue with this module for now, and decide in 6 months / 1 year if we need to change something. WDYT?

benoit74 commented on api/src/zimitfrontend/__about__.py:
```diff
@@ -0,0 +1,5 @@
+__version__ = "0.0.0-dev0"
```
Nope: https://github.com/openzim/_python-bootstrap/blob/main/src/great_project/__about__.py (I don't remember exactly why, but we've settled on `-dev0` after some discussions.

Regarding `0.0.0`, it just means "not versioned", I feel like it is way more explicit than living with `0.0.1-dev0` forever. Do you recommend to use `0.0.1-dev0`?

benoit74 commented on api/src/zimitfrontend/constants.py:
```diff
@@ -0,0 +1,97 @@
+    )
+    hook_token = os.getenv("HOOK_TOKEN", uuid.uuid4().hex)
+
+    ui_location = pathlib.Path(os.getenv("UI_LOCATION", "/src/ui"))
```
Do I get it correctly that it is not needed to provide an empty default value, just no default is OK?

benoit74 replied:
forget about it, I wasn't on proper line ...

benoit74 commented on api/src/zimitfrontend/routes/utils.py:
```diff
@@ -0,0 +1,129 @@
+    }
+
+    logger.info(f"Translating to {lang}")
+    set_locale(lang)
```
I renamed it to `change_locale`. Idea is to isolate `python-i18n` in `i18n.py` module so that it is easy to replace.

benoit74 commented on api/src/zimitfrontend/utils.py:
```diff
@@ -0,0 +1,65 @@
+jinja_env.globals["translate"] = i18n.t
+
+
+def send_email_via_mailgun(
```
Fixed

benoit74 commented:
> I tested it and it seems to work fine. I was somehow surprised that the UI looks the same 😀

Mostly 🤫

> On the python side, I think there are a couple of place with poor type hints (zimfarm module?).

Fixed, thank you

> On the Vue part, I mostly skimmed through the code but it looks good and coherent.

OK, thank you

> On personal taste, I find the user feedback too dim. If there is an error in making a request, theres just a light toast at the bottom of the page that disappears after 5s.

I think you refer to home page issue, when it fails to fetch offliner definition. This is indeed something I forgot about. I changed the toast to red + I added a message in red as well which stays forever. On the page about the task status, there is a red message when we fail to fetch the task.

benoit74 commented:
(merging because confirmed with rgaudin that further review is not necessary once conversations are taken care of)
